### ISSUE TITLE
Performance test errors in CI

### ISSUE THREAD

**Issue by vdimir** [ClickHouse/ClickHouse#66978]
Performance test are failing regularly with errors.
Some failures related to server being killed by OOM killer.

See [the list of errors](https://pastila.nl/?0000d433/1ec68fc1ab743effee4b3997efbf6b31#NDn5u9jzGQWTLAxZQjhg9A==), `has_oom` column indicates if the error was caused by OOM killer and filled with a [script](https://pastila.nl/?05b1030e/a7382de8953c89a2addaf97a946c5308#j5nYJbPJ4GRY8c8ZuPqnTg==), checking `dmest.log` `output.7z`

[Example dmesg log](https://pastila.nl/?0033bb09/e257b47cc405bf17ab9c26f65d6c9192#2R5qUVvjdS84no8JFZaEtw==)

`clickhouse-server` consumed around 15GB RSS each, while total memory is 60GB, so it's not clear why OOM killer was triggered.

Continuing investigation ...

**Comment by antonio2368** [comment]
Maybe it's the fact we bind to NUMA node which has a lower limit

From your dmesg
```
Node 1 Normal free:50072kB boost:0kB min:45152kB low:77452kB high:109752kB reserved_highatomic:2048KB active_anon:29986116kB inactive_anon:1120612kB active_file:0kB inactive_file:5800kB unevictable:600288kB writepending:1284kB present:32821248kB managed:32301916kB mlocked:600288kB bounce:0kB free_pcp:396kB local_pcp:0kB free_cma:0kB
```

`free` is less than `low` and from my basic understanding of such things and quick research that should trigger OOM killer.

@vdimir can you give me the link to the run itself?

**Comment by antonio2368** [comment]
Also `active_anon:29986116kB` which is around 30GB which is exactly how much 2 servers combined took.
It's not clear to me yet what we want to do here, do we want to use the limit of the NUMA node instead of the entire system? :thinking:

**Comment by vdimir** [comment]
@antonio2368 linked dmesg is from [this report](https://s3.amazonaws.com/clickhouse-test-reports/0/4f02ded96e6896e303ec5752b24fc06855657cfb/performance_comparison__release__[1_4]/report.html). 

But actually, I checked several entries marked with `oom` in the list from the first comment, and they all have similar symptoms.

**Comment by antonio2368** [comment]
Okay it bound to the node 1 which is problematic.
I think we can try to do something with https://linux.die.net/man/3/numa
E.g. use `numa_get_membind` and `numa_node_size` and set memory limit based on that

**Comment by vdimir** [comment]
Can we not bound process (probably it'll lead to unstable result)? If we lower max_memory_usage, then it'll be MEMORY_LIMIT_EXCEPTION instead oom in the same tests, or how it can help mitigate the memory shortage?

**Comment by antonio2368** [comment]
> Can we not bound process (probably it'll lead to unstable result)? 

Yes, ofc this will fix the issue.
I'm just checking if there is a way to fix such things generally.

**Comment by antaljanosbenjamin** [comment]
> Yes, ofc this will fix the issue.

That is true, however it would introduce some uncertanity to the perf checks. I have no idea how much does it matter, but it might worth to closely measure that.

**Comment by antonio2368** [comment]
I managed to build the numa library and get information about the numa nodes (the hardest part was updating sysroot)
We can easily extend `getMemoryAmount` to detect the case if a process is bound to a single numa node and get its size
But what if multiple nodes are allowed? If each node can have its own limit it would be impossible for us to correctly limit the memory.

**Comment by antaljanosbenjamin** [comment]
> But what if multiple nodes are allowed? If each node can have its own limit it would be impossible for us to correctly limit the memory.

If we can improve the case when CH is bound to a single numa node, then it is already an improvement, therefore I would say it make sense to do it.

For the case when CH is bound to multiple numa nodes, I don't know any good solution, but so far we don't use CH in such a way and nobody reported issues regarding that. So maybe we can ignore it (until somebody pops up).

**Comment by antonio2368** [comment]
https://github.com/ClickHouse/ClickHouse/pull/67098 to at least try detecting correct total memory.

**Comment by azat** [comment]
I haven't seen OOM in perf tests for a long time. Let's close for now.
